### PR TITLE
Fasta-stats: replace perl by python 

### PR DESCRIPTION
**Pull request by gallardoalba**
FOR CONTRIBUTOR:
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [ ] - This PR adds a new tool or tool collection
* [ ] - This PR updates an existing tool or tool collection
* [x] - This PR does something else (explain below)

This PR updates the FASTA-stats tool. Concretely, it replaces the previous Perl script by a Python version developed during the BioHackathon.

### CODE REVIEW

File: tools/fasta_stats/fasta-stats.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+# git: @codemeleon
+# date: 10/11/2021
+
+
+import click
+import numpy as np
+from tabulate import tabulate
+from Bio import SeqIO
+import re
+import pandas as pd
```
Comment (bgruening) [inline tools/fasta_stats/fasta-stats.py:15]: @codemeleon can we drop the pandas dependency? Its a heavy dependency and we do not need it, isn't it? We can simply write the values as we calculate them, line by line.

File: tools/fasta_stats/fasta-stats.xml [hunk 2]
```diff
@@ -1,66 +1,75 @@
-<tool id="fasta-stats" name="Fasta Statistics" version="1.0.3">
-    <description>Display summary statistics for a fasta file.</description>
+<tool id="fasta-stats" name="Fasta Statistics" version="1.0" profile="20.05">
```
Comment (bgruening) [inline tools/fasta_stats/fasta-stats.xml:1]: @gallardoalba we need to increase the version. It needs to be bigger than 1.0.3

File: tools/fasta_stats/fasta-stats.xml [hunk 3]
```diff
@@ -1,66 +1,75 @@
...
+        <data name="gap_stats" format="bed" label="${tool.name} on ${on_string}: Gap stats">
+            <filter>gaps_option</filter>
+        </data>
     </outputs>
     <tests>
-        <test>
-            <param name="dataset" value="test.fasta"/>
-            <output name="stats" file="test_out.txt"/>
+        <test expect_num_outputs="1">
+            <param name="input_fasta" value="scaffold.fasta"/>
```
Comment (bernt-matthias) [inline tools/fasta_stats/fasta-stats.xml:29]: Keeping the old test data would have the advantage that we see that the output is identical.

The test data seems quite large.

**Comment by Slugger70** [comment]
Why are the NG50 and LG50 stats being dropped? They were requested a while back by a user.

File: tools/fasta_stats/fasta-stats.py (hunk 1 above)
Comment (codemeleon) [reply tools/fasta_stats/fasta-stats.py:15]: Sure. I will convert the code accordingly.

Comment (codemeleon) [reply tools/fasta_stats/fasta-stats.py:15]: I have uploaded updated file at https://raw.githubusercontent.com/codemeleon/BitterBits/master/src/galaxy/fasta-stats.py

Comment (codemeleon) [reply tools/fasta_stats/fasta-stats.py:15]: I have uploaded updated file at https://raw.githubusercontent.com/codemeleon/BitterBits/master/src/galaxy/fasta-stats.py . Please let me know if code is OK.

Comment (bgruening) [reply tools/fasta_stats/fasta-stats.py]: Perfect! I added it!

Comment (bgruening) [reply tools/fasta_stats/fasta-stats.py]: @codemeleon do you have any plans to add NG50 and LG50? That was previously requested by the team. Is that something you can work on or should we?

Comment (codemeleon) [reply tools/fasta_stats/fasta-stats.py]: I update you after including them in the code.

File: tools/fasta_stats/fasta-stats.xml [hunk 4]
```diff
@@ -1,66 +1,75 @@
...
-    <description>Display summary statistics for a fasta file.</description>
+<tool id="fasta-stats" name="Fasta Statistics" version="1.0" profile="20.05">
+    <description>display summary statistics for a FASTA file</description>
     <requirements>
-        <requirement type="package" version="5.26">perl</requirement>
+        <requirement type="package" version="0.8.9">tabulate</requirement>
+        <requirement type="package" version="8.0.3">click</requirement>
+        <requirement type="package" version="1.21.4">numpy</requirement>
+        <requirement type="package" version="1.3.4">pandas</requirement>
+        <requirement type="package" version="1.79">biopython</requirement>
```
Comment (wm75) [inline tools/fasta_stats/fasta-stats.xml:8]: Can this be cleaned up a bit?
I don't think tabulate and pandas being used anywhere. You could remove click if you used argparse instead (really no need to do anything special about a cli that nobody will ever see exposed).

File: tools/fasta_stats/fasta-stats.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    scaffold_lens.sort(reverse=True)
+    scaffold_lens = np.array(scaffold_lens)
+    scaffold_lens = np.cumsum(scaffold_lens)
+
+    n50_len = scaffold_lens[-1] * 0.5
+    n50_idx = np.where(scaffold_lens > n50_len)[0][0]
+    n90_len = scaffold_lens[-1] * 0.9
+    n90_idx = np.where(scaffold_lens > n90_len)[0][0]
+
+    to_csv = []
```
Comment (wm75) [inline tools/fasta_stats/fasta-stats.py:111]: very misleading name for something that gets turned into tab-separated in the end

File: tools/fasta_stats/fasta-stats.xml (hunk 4 above)
Comment (bgruening) [reply tools/fasta_stats/fasta-stats.xml:8]: @wm75 we just removed this dependency from the python code 5min ago :)

Comment (codemeleon) [reply tools/fasta_stats/fasta-stats.xml:8]: @wm75 Do want me to remove click??

File: tools/fasta_stats/fasta-stats.py (hunk 5 above)
Comment (codemeleon) [reply tools/fasta_stats/fasta-stats.py:111]: @wm75 Could you please elaborate?

File: tools/fasta_stats/fasta-stats.py (hunk 1 above)
Comment (codemeleon) [reply tools/fasta_stats/fasta-stats.py]: Done.

Comment (codemeleon) [reply tools/fasta_stats/fasta-stats.py]: Done.

File: tools/fasta_stats/fasta-stats.xml (hunk 4 above)
Comment (wm75) [reply tools/fasta_stats/fasta-stats.xml:8]: Not a must, but having less dependencies is always good :)

Comment (codemeleon) [reply tools/fasta_stats/fasta-stats.xml:8]: I will keep in mind. Thank you.

File: tools/fasta_stats/fasta-stats.py (hunk 5 above)
Comment (wm75) [reply tools/fasta_stats/fasta-stats.py:111]: Well, you're never writing comma-separated (csv) output, do you? Same for gaps_csv.
Sure, the code doesn't care, but it's misleading for the reader.

Comment (codemeleon) [reply tools/fasta_stats/fasta-stats.py:111]: Yes. It does write output tsv file,if you provide output file names. Else it outputs as stdout.

Comment (gallardoalba) [reply tools/fasta_stats/fasta-stats.py:111]: If you prefer @codemeleon I can close this PR and you can open a new one with the modifications.

Comment (codemeleon) [reply tools/fasta_stats/fasta-stats.py:111]: @gallardoalba . Sure. I hope, all issues are resolved.

**Comment by gallardoalba** [comment]
> Why are the NG50 and LG50 stats being dropped? They were requested a while back by a user.

It has been included again.

**Comment by gallardoalba** [comment]
Hi @codemeleon, I included the modifications of your code. You did a great job.

File: tools/fasta_stats/fasta-stats.xml [hunk 6]
```diff
@@ -1,66 +1,78 @@
...
     </requirements>
     <command detect_errors="exit_code"><![CDATA[
-        perl '${__tool_directory__}/fasta-stats.pl'
-        '$dataset'
-        #if $genome_size:
-            $genome_size
+        python '${__tool_directory__}/fasta-stats.py'
+        '$input_fasta'
+        '$stats'
+        '$gap_stats'
```
Comment (wm75) [inline tools/fasta_stats/fasta-stats.xml:11]: Think this should be skipped when str($gaps_option) == 'false'?

File: tools/fasta_stats/fasta-stats.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+                continue
+            for rng in seq_id_Ngaprange[key]:
+                goutput.write("{}\t{}\t{}\n".format(key, rng[0], rng[1]))
+
+
+if __name__ == "__main__":
+
+    fasta_file = argv[1]
+    stats_output = argv[2]
+    gaps_output = argv[3]
```
Comment (wm75) [inline tools/fasta_stats/fasta-stats.py:179]: Make gaps_output optional?

File: tools/fasta_stats/fasta-stats.xml [hunk 8]
```diff
@@ -1,66 +1,78 @@
...
+        <!--Test gap options and NG50-->
+        <test expect_num_outputs="2">
+            <param name="input_fasta" value="ng50_input.fasta"/>
+            <param name="gaps_option" value="true"/>
+            <param name="estimated_genome_size" value="4000"/>
+            <output name="stats" file="test_02.tab" ftype="tabular"/>
+            <output name="gap_stats" file="test_02.bed" ftype="bed"/>
         </test>
     </tests>
     <help>
```
Comment (wm75) [inline tools/fasta_stats/fasta-stats.xml:42]: use CDATA for help section

File: tools/fasta_stats/fasta-stats.xml [hunk 9]
```diff
@@ -1,66 +1,78 @@
-<tool id="fasta-stats" name="Fasta Statistics" version="1.0.3">
-    <description>Display summary statistics for a fasta file.</description>
+<tool id="fasta-stats" name="Fasta Statistics" version="1.0.4" profile="20.05">
```
Comment (wm75) [inline tools/fasta_stats/fasta-stats.xml:1]: I'd say, a complete rewrite of a tool plus added new functionality warrants a 2.0 version?

File: tools/fasta_stats/fasta-stats.py [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+        )
+        soutput.write("{}\t{}\n".format("Contig len_std", int(np.std(seq_len_list))))
+        soutput.write("{}\t{}\n".format("Contig num_bp", contigs_len_sum[-1]))
+        soutput.write("{}\t{}\n".format("Contig num_seq", len(contigs_len_sum)))
+
+    # NOTE: generate gaps statistics file
+    with open(gaps_output, "w") as goutput:
+        for key in seq_id_Ngaprange:
+            if not len(seq_id_Ngaprange[key]):
+                continue
```
Comment (wm75) [inline tools/fasta_stats/fasta-stats.py:170]: Can be dropped. Cause the following for loop will just not do anything on an empty iterable.

File: tools/fasta_stats/fasta-stats.xml (hunk 6 above)
Comment (wm75) [reply tools/fasta_stats/fasta-stats.xml:11]: any idea why the first test works and what $gap_stats is set to when gaps_option is false?

Comment (gallardoalba) [reply tools/fasta_stats/fasta-stats.xml:11]: $gap_stats is only the output dataset, so when gap_options is false, it is generated, but not shown. I'll modify  that.

Comment (gallardoalba) [reply tools/fasta_stats/fasta-stats.xml:11]: $gap_stats is only the output dataset, so when gap_options is false, it is generated, but not shown. I'll modify  that.

File: tools/fasta_stats/test-data/test_01.tab [hunk 11]
```diff
@@ -16,7 +16,6 @@ Scaffold num_bp	194780
 Scaffold num_bp_not_N	194780
 Scaffold num_seq	1
 Scaffold GC content overall	52.01
-Number of gaps	0
```
Comment (bgruening) [inline tools/fasta_stats/test-data/test_01.tab]: To be honest, I would rather leave the 0. This way the param-from-file can get a value at any time. ping @wm75

Comment (gallardoalba) [reply tools/fasta_stats/test-data/test_01.tab]: Sure. I'll include it again.

File: tools/fasta_stats/fasta-stats.py [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+        required=True,
+        help="File to store the general statistics",
+    )
+    parser.add_argument(
+        "-r",
+        "--gaps_output",
+        required=False,
+        help="File to store the gaps statistics",
+        default=None,
+    )
```
Comment (wm75) [inline tools/fasta_stats/fasta-stats.py:203]: This option is all that's needed, isn't it, and the `--gaps_option` flag is just redundant. If args.gaps_output is None, you don't produce the extra file.

File: tools/fasta_stats/fasta-stats.xml [hunk 13]
```diff
@@ -1,66 +1,81 @@
...
-        '$dataset'
-        #if $genome_size:
-            $genome_size
+        python '${__tool_directory__}/fasta-stats.py'
+        --fasta '$fasta'
+        --stats_output '$stats_output'
+        --gaps_option $gaps_option
+        #if $gaps_option
+            --gaps_output '$gaps_output'
+        #end if
```
Comment (wm75) [inline tools/fasta_stats/fasta-stats.xml:14]: as in the python script, this is all that's needed. No need to pass $gaps_option itself.

**Review by wm75 — DISMISSED** [review]

**Comment by gallardoalba** [comment]
There are some inconsistencies in the script when compared with QUAST ([Galaxy history](https://usegalaxy.eu/u/gallardoalba/h/assemblystats))

Screenshot from 2021-11-18 10-05-09
:

**Comment by wm75** [comment]
@gallardoalba I think that's because for all the N50, N90 calculations the code indexes into the original SEQ_LEN_LIST/seq_len_list lists, but these are *unsorted*.

**Comment by wm75** [comment]
So I guess if you'd sort *before* assigning to `SEQ_LEN_LIST` and `seq_len_list`, you'd be fine.
In addition, that would simplify other calculations:
- I'm sure np.median needs to sort internally anyway,
- np.max and np.min could be replaced with simply last and first element of the lists
and I don't see the unsorted list being required anywhere.

**Comment by gallardoalba** [comment]
> @gallardoalba I think that's because for all the N50, N90 calculations the code indexes into the original SEQ_LEN_LIST/seq_len_list lists, but these are _unsorted_.

Yes, you were totally right. I included the modifications in the code.

File: tools/fasta_stats/fasta-stats.py [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+                    contigs_len.append(con_len)
+        else:
+            contigs_len.append(len(seq))
+
+    SEQ_LEN_LIST = sorted(list(seq_len.values()), reverse=True)
+    scaffold_lens = list(SEQ_LEN_LIST)
+
+    # NOTE: Scaffold statistics
+    scaffold_lens.sort(reverse=True)
+    scaffold_lens = np.array(scaffold_lens)
```
Comment (wm75) [inline tools/fasta_stats/fasta-stats.py:92]:
```suggestion
    # NOTE: Scaffold statistics
    SEQ_LEN_LIST = sorted(seq_len.values(), reverse=True)
    scaffold_lens = np.array(SEQ_LEN_LIST)
```

File: tools/fasta_stats/fasta-stats.py [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+    N50_len = scaffold_lens_sum[-1] * 0.5
+    N50_idx = np.where(scaffold_lens_sum > N50_len)[0][0]
+    N90_len = scaffold_lens_sum[-1] * 0.9
+    N90_idx = np.where(scaffold_lens_sum > N90_len)[0][0]
+    NG50 = calculate_NG50(genome_size, scaffold_lens_sum[-1], scaffold_lens)
+
+    # NOTE: Contig statistics
+    contigs_len.sort(reverse=True)
+    seq_len_list = list(contigs_len)
+    contigs_len = np.array(contigs_len)
```
Comment (wm75) [inline tools/fasta_stats/fasta-stats.py:103]:
```suggestion
    seq_len_list = sorted(contigs_len, reverse=True)
    contigs_len = np.array(seq_len_list)
```

File: tools/fasta_stats/fasta-stats.py [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+            )
+        )
+        soutput.write("{}\t{}\n".format("Scaffold num_seq", len(SEQ_LEN_LIST)))
+        soutput.write(
+            "{}\t{:.2f}\n".format(
+                "Scaffold GC content overall",
+                (
+                    (bases_global["G"] + bases_global["C"])
+                    * 100.0
+                    / sum(seq_len.values())
```
Comment (wm75) [inline tools/fasta_stats/fasta-stats.py:145]:
```suggestion
                    / scaffold_lens_sum[-1]
```

**Review by wm75 — COMMENTED** [review]
Suggested a bit of a cleanup of the lists zoo

**Review by wm75 — APPROVED** [review]
LGTM (again).
Thanks a lot @gallardoalba @codemeleon !

**Comment by gallardoalba** [comment]
> LGTM (again). Thanks a lot @gallardoalba @codemeleon !

@Mpalmada confirmed to me that the problem has been fixed.

**Comment by bgruening** [comment]
Merci @gallardoalba @codemeleon!

**Comment by codemeleon** [comment]
> uess if you'd sort _before_ assigning to `SEQ_LEN_LIST` and `seq_len_list`, you'd be fine.



>     * np.median needs to sort internally anyway,
yes. You are correct. 


> So I guess if you'd sort _before_ assigning to `SEQ_LEN_LIST` and `seq_len_list`, you'd be fine. In addition, that would simplify other calculations:
> 
>     * I'm sure np.median needs to sort internally anyway,
> 
>     * np.max and np.min could be replaced with simply last and first element of the lists
>       and I don't see the unsorted list being required anywhere.
> 
> 
> @codemeleon you know the code best. Am I talking nonsense here?

Max will be the first element and Min will be the last, as the list will be sorted in decreasing order.

**Pull request merged**